File: autoburn.pc/autoburn/msghandler/RxMsgDispatch.cs
using Autoburn.util;
using Newtonsoft.Json.Linq;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Autoburn.Net;
using Autoburn.Manager;

namespace Autoburn.MsgHandler
{
    public class RxMsgDispatch
    {
        private string TAG = "RxMsgDispatch";

        private HashSet<string> _SupportRxMsg = new HashSet<string>();
        private DeviceManager deviceManager;

        private void AddHandler()
        {
            _SupportRxMsg.Clear();
            _SupportRxMsg.Add(new RxMsgHandlerInfo(deviceManager).Type);
            _SupportRxMsg.Add(new RxMsgHandlerTest(deviceManager).Type);
        }

        private RxMsgDispatch()
        {
            AddHandler();
        }

        public RxMsgDispatch(DeviceManager deviceManager) : this()
        {
            this.deviceManager = deviceManager;
        }

        public void Stop()
        {
            _SupportRxMsg?.Clear();
        }

        public void ProcessRxMsg(string str)
        {
            if (string.IsNullOrEmpty(str))
            {
                return;
            }
            RxMsgHandlerBase rxMsgHandler = null;
            try
            {
                JObject obj = JObject.Parse(str);
                string msgtype = obj[MsgBase.MSG_TYPE_STRING].ToString();
                if (!_SupportRxMsg.Contains(msgtype))
                {
                    //   return;
                }
                ProgLog.D(TAG, "the msgtype is " + msgtype);
                switch (msgtype)
                {
                    case MsgBase.MSG_TYPE_INFO:
                        rxMsgHandler = new RxMsgHandlerInfo(deviceManager);
                        break;
                    case MsgBase.MSG_TYPE_TEST:
                        rxMsgHandler = new RxMsgHandlerTest(deviceManager);
                        break;
                    default:
                        break;
                }

                if (rxMsgHandler != null)
                {
                    rxMsgHandler.Msg = str;
                }
            }
            catch (Exception e)
            {
                ProgLog.D(TAG, "ProcessRxMsg error: " + e.ToString());
            }

            if (rxMsgHandler != null)
            {
                Task.Run(new Action(rxMsgHandler.Process));
            }
        }
    }
}

File: autoburn.android/app/src/main/java/com/autoburn/msghandler/RxMsgDispatch.java
package com.autoburn.msghandler;

import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashSet;

/**
 * Created by chowhsienwu on 2017/5/4.
 */
public class RxMsgDispatch {
    public static final String TAG = "RxMsgDispatch";

    public RxMsgDispatch(){
        addHandler();
    }

    private void addHandler(){
        mSupportRxMsg.clear();
        mSupportRxMsg.add(new RxMsgHandlerInfo().Type);
        mSupportRxMsg.add(new RxMsgHandlerTest().Type);
    }

    private HashSet<String> mSupportRxMsg= new HashSet<String>();

    public void stop(){
        mSupportRxMsg.clear();
    }

    public void processRxMsg(String str){
        if (str == null || str.length() < 1){
            return;
        }
        RxMsgHandlerBase rxMsgHandlerBase = null;
        try {
            JSONObject jo = new JSONObject(str);
            String msgType = null;
            msgType = jo.getString(MsgBase.MSG_TYPE_STRING);
            Log.i(TAG, "the json type is  " + msgType);
            switch (msgType) {
                case MsgBase.MSG_TYPE_INFO:
                    rxMsgHandlerBase = new RxMsgHandlerInfo();
                    break;
                case MsgBase.MSG_TYPE_TEST:
                    rxMsgHandlerBase = new RxMsgHandlerTest();
                default:
                    break;
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }

        if (rxMsgHandlerBase != null){
            // parse the msg in a new asynctask
            rxMsgHandlerBase.execute(str);
        }

    }

}
